entity/content_context.cc

```diff
@@ -24,24 +24,41 @@ ContentContext::ContentContext(std::shared_ptr<Context> context)
 
   // Pipelines that are variants of the base pipelines with custom descriptors.
   if (auto solid_fill_pipeline = solid_fill_pipelines_[{}]->WaitAndGet()) {
-    auto clip_pipeline_descriptor = solid_fill_pipeline->GetDescriptor();
-    clip_pipeline_descriptor.SetLabel("Clip Pipeline");
-    // Write to the stencil buffer.
-    StencilAttachmentDescriptor stencil0;
-    stencil0.stencil_compare = CompareFunction::kGreaterEqual;
-    stencil0.depth_stencil_pass = StencilOperation::kSetToReferenceValue;
-    clip_pipeline_descriptor.SetStencilAttachmentDescriptors(stencil0);
-    // Disable write to all color attachments.
-    auto color_attachments =
-        clip_pipeline_descriptor.GetColorAttachmentDescriptors();
-    for (auto& color_attachment : color_attachments) {
-      color_attachment.second.write_mask =
-          static_cast<uint64_t>(ColorWriteMask::kNone);
+    // Clip pipeline.
+    {
+      auto clip_pipeline_descriptor = solid_fill_pipeline->GetDescriptor();
+      clip_pipeline_descriptor.SetLabel("Clip Pipeline");
+      // Write to the stencil buffer.
+      StencilAttachmentDescriptor stencil0;
+      stencil0.stencil_compare = CompareFunction::kEqual;
+      stencil0.depth_stencil_pass = StencilOperation::kIncrementClamp;
+      clip_pipeline_descriptor.SetStencilAttachmentDescriptors(stencil0);
+      // Disable write to all color attachments.
+      auto color_attachments =
+          clip_pipeline_descriptor.GetColorAttachmentDescriptors();
+      for (auto& color_attachment : color_attachments) {
+        color_attachment.second.write_mask =
+            static_cast<uint64_t>(ColorWriteMask::kNone);
+      }
+      clip_pipeline_descriptor.SetColorAttachmentDescriptors(
+          std::move(color_attachments));
+      clip_pipelines_[{}] =
+          std::make_unique<ClipPipeline>(*context_, clip_pipeline_descriptor);
+    }
+
+    // Clip restoration pipeline.
+    {
+      auto clip_pipeline_descriptor =
+          clip_pipelines_[{}]->WaitAndGet()->GetDescriptor();
```

Comment:
Just a drive by comment that you absolutely should not account for here because the right™ [fix is account for the issue with having to wait for a pipeline future just to get its descriptor](https://github.com/flutter/impeller/blob/0ad43eac6fa00bdbb68d7538b3dc53f4b7a66d8b/renderer/pipeline.h#L18) it to rework pipeline futures.

A restore pipeline is a variant of the clip pipeline which is a variant of the solid fill pipeline. And you have already awaited for the fill pipeline so you already have the descriptor. You could reuse that descriptor directly instead of awaiting the full initialization of the clip pipeline to prepare the restore variant.

Either way though, this is correct and an issue with the API forcing an await for a descriptor when it should only do so for the actual pipeline itself.

Reply:
Consider filing an issue and writing a TODO for follow-up work.

Reply:
I filed https://github.com/flutter/flutter/issues/98684 specifically for this issue. Backfilling the issues in the bug tracker now (adding the impeller label). Will update the TODOs to reference the bugs once all issues are filed.